src/python/pants/build_graph/address.py

```diff
@@ -325,44 +338,61 @@ def spec(self) -> str:
                 else f"{parent_prefix}{self.target_name}"
             )
         target_sep = ":" if target else ""
+        if self.parameters:
+            params_sep = "@"
+            params = ",".join(f"{k}={v}" for k, v in self.parameters.items())
+        else:
+            params_sep = ""
+            params = ""
         if self.generated_name is None:
             generated_sep = ""
             generated = ""
         else:
             generated_sep = "#"
             generated = self.generated_name
 
-        return f"{prefix}{path}{target_sep}{target}{generated_sep}{generated}"
+        return f"{prefix}{path}{target_sep}{target}{params_sep}{params}{generated_sep}{generated}"
 
     @property
     def path_safe_spec(self) -> str:
         """
         :API: public
         """
+
+        def sanitize(s: str) -> str:
+            return s.replace(os.path.sep, ".")
+
         if self._relative_file_path:
             parent_count = self._relative_file_path.count(os.path.sep)
             parent_prefix = "@" * parent_count if parent_count else "."
-            path = f".{self._relative_file_path.replace(os.path.sep, '.')}"
+            path = f".{sanitize(self._relative_file_path)}"
         else:
             parent_prefix = "."
             path = ""
         if parent_prefix == ".":
             target = f"{parent_prefix}{self._target_name}" if self._target_name else ""
         else:
             target = f"{parent_prefix}{self.target_name}"
-        generated = (
-            f"@{self.generated_name.replace(os.path.sep, '.')}" if self.generated_name else ""
-        )
-        prefix = self.spec_path.replace(os.path.sep, ".")
-        return f"{prefix}{path}{target}{generated}"
+        if self.parameters:
+            key_value_strs = ",".join(
+                f"{sanitize(k)}={sanitize(v)}" for k, v in self.parameters.items()
+            )
+            params = f"@@{key_value_strs}"
```

Comment:
Why is the double @ necessary for the path-safe spec? It is not immediately obvious to me.

Reply:
Um... me neither. Honestly, they could probably use a re-think. But my thinking was that we use `@` to escape other positions in the `path_safe_spec`, and so doubling it ... reduced the chance of collision with that component?